src/OpenCLScheduler.cpp

```diff
@@ -22,61 +22,75 @@
 #include "Random.h"
 #include "OpenCLScheduler.h"
 
-thread_local auto current_thread_gpu_num = size_t{0};
-OpenCLScheduler opencl;
 
 void OpenCLScheduler::initialize(const int channels) {
     // multi-gpu?
-    if (!cfg_gpus.empty()) {
-        auto silent{false};
-        for (auto gpu : cfg_gpus) {
-            auto opencl = std::make_unique<OpenCL>();
-            auto net = std::make_unique<OpenCL_Network>(*opencl);
-            opencl->initialize(channels, {gpu}, silent);
-            m_opencl.push_back(std::move(opencl));
-            m_networks.push_back(std::move(net));
+    auto gpus = cfg_gpus;
+    if (gpus.empty()) {
+        gpus = {0};
+    }
 
-            // Clear thread data on every init call.  We don't know which GPU
-            // this thread will be eventually be assigned to
-            opencl_thread_data = ThreadData();
+    auto silent{false};
+    auto gnum = size_t{0};
 
-            // starting next GPU, let's not dump full list of GPUs
-            silent = true;
-        }
+    // launch the worker thread.  2 threads so that we can fully
+    // utilize GPU, since the worker thread consists of some CPU
+    // work for task preparation.
+    constexpr auto num_threads = 2;
+    m_context.resize(num_threads);
 
-        for (size_t gnum = 0; gnum < m_networks.size(); gnum++) {
-            // launch the worker thread.  2 threads so that we can fully
-            // utilize GPU, since the worker thread consists of some CPU
-            // work for task preparation.
-            constexpr auto num_threads = 2;
-            for (auto i = 0; i < num_threads; i++) {
-                m_threadpool.add_thread([gnum] {
-                    current_thread_gpu_num = gnum;
-                });
-            }
-        }
-    } else {
+    for (auto gpu : gpus) {
         auto opencl = std::make_unique<OpenCL>();
         auto net = std::make_unique<OpenCL_Network>(*opencl);
-        opencl->initialize(channels, {});
-
+        opencl->initialize(channels, {gpu}, silent);
         m_opencl.push_back(std::move(opencl));
         m_networks.push_back(std::move(net));
```

Comment:
Does this work correctly with --gpu 1 --gpu 3?

m_networks will be a size 2 array, yet ctx->m_gpu_num which is indexing into that can be 1 and 3.

Reply:
Ouch x100.  Will rewrite code.

Reply:
While trying to rewrite code, figured out that this isn't a bug - it is correct because m_gpu_num will be 0 or 1 (that is, this isn't the GPU number given by the command line argument.  This is the GPU index number on OpenCLScheduler.  Looking back, I don't like defining variables that are only used on OpenCLScheduler on a OpenCL data structure.  Will do a small rewrite to make it prettier...